### PR TITLE
perf: Parallelize read calls by table and batch

### PR DESCRIPTION
**Pull request by robhowley**
# What this PR does / why we need it:

Improve the use of async batch calls to dynamo db for `get_online_features_async`

* tested on a feature service w 3 feature views and 11 features
* compared to the sync method, the async method was
  * 50% faster w 5 entities retrieved
  * 75% faster w 80 entities retrieved

# Which issue(s) this PR fixes:

# Misc

### CODE REVIEW

File: sdk/python/feast/infra/online_stores/dynamodb.py [hunk 1]
```diff
@@ -309,24 +309,40 @@ def to_tbl_resp(raw_client_response):
...
+        entity_id_batches = []
+        while True:
+            batch = list(itertools.islice(entity_ids_iter, batch_size))
+            if not batch:
+                break
+            entity_id_batch = self._to_client_batch_get_payload(
+                online_config, table_name, batch
+            )
+            batches.append(batch)
+            entity_id_batches.append(entity_id_batch)
```
Comment (robhowley) [inline sdk/python/feast/infra/online_stores/dynamodb.py:322]: construct the batches of ids/entity_ids that we'll be looking up

File: sdk/python/feast/infra/online_stores/dynamodb.py [hunk 2]
```diff
@@ -309,24 +309,40 @@ def to_tbl_resp(raw_client_response):
...
-                result.extend(batch_result)
-        return result
+            response_batches = await asyncio.gather(
+                *[
+                    client.batch_get_item(
+                        RequestItems=entity_id_batch,
+                    )
+                    for entity_id_batch in entity_id_batches
+                ]
+            )
```
Comment (robhowley) [inline sdk/python/feast/infra/online_stores/dynamodb.py:332]: make those batch requests in parallel. 

_note_: `gather` maintains order

File: sdk/python/feast/infra/online_stores/dynamodb.py [hunk 3]
```diff
@@ -309,24 +309,40 @@ def to_tbl_resp(raw_client_response):
...
+        result_batches = []
+        for batch, response in zip(batches, response_batches):
+            result_batch = self._process_batch_get_response(
+                table_name,
+                response,
+                entity_ids,
+                batch,
+                to_tbl_response=to_tbl_resp,
+            )
+            result_batches.append(result_batch)
```
Comment (robhowley) [inline sdk/python/feast/infra/online_stores/dynamodb.py:343]: format the responses to the final format. we iterate through the list three times in stead of one, but make up for it in asyncing the batches

File: sdk/python/feast/infra/online_stores/online_store.py [hunk 4]
```diff
@@ -258,6 +258,18 @@ async def get_online_features_async(
...
+        all_responses = await asyncio.gather(
+            *[
+                query_table(table, requested_features)
+                for table, requested_features in grouped_refs
+            ]
+        )
+
+        for (idxs, read_rows), (table, requested_features) in zip(
+            all_responses, grouped_refs
+        ):
```
Comment (robhowley) [inline sdk/python/feast/infra/online_stores/online_store.py:272]: when requesting features across multiple tables, we can parallelize the calls to each.

File: sdk/python/feast/infra/online_stores/online_store.py [hunk 5]
```diff
@@ -240,7 +240,7 @@ async def get_online_features_async(
             native_entity_values=True,
         )
 
-        for table, requested_features in grouped_refs:
+        async def query_table(table, requested_features):
```
Comment (franciscojavierarceo) [inline sdk/python/feast/infra/online_stores/online_store.py:243]: can you add type hints?

**Comment by franciscojavierarceo** [comment]
lgtm

**Review by franciscojavierarceo — APPROVED** [review]

**Pull request merged**